File: central/app/services/hooks.py
from __future__ import annotations

import os
import subprocess
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


MAX_HOOK_FILES = 3
_ALLOWED_SUFFIXES = {"", ".sh", ".txt"}
_HOOK_TIMEOUT_SECONDS = 300


class HookManager:
    def __init__(self, scripts_dir: Path, logger) -> None:
        self.scripts_dir = scripts_dir
        self.logger = logger
        self.scripts_dir.mkdir(parents=True, exist_ok=True)

    def snapshot(self, *, pre_command: str, post_command: str) -> dict[str, Any]:
        return {
            "pre_command": pre_command,
            "post_command": post_command,
            "script_dir": str(self.scripts_dir),
            "max_files": MAX_HOOK_FILES,
            "files": self.list_files(),
        }

    def list_files(self) -> list[dict[str, Any]]:
        self.scripts_dir.mkdir(parents=True, exist_ok=True)
        files = sorted((path for path in self.scripts_dir.iterdir() if path.is_file()), key=lambda path: path.name.lower())
        return [self._file_payload(path) for path in files[:MAX_HOOK_FILES]]

    def save_uploaded_file(self, filename: str, content: bytes) -> dict[str, Any]:
        safe_name = self._sanitize_filename(filename)
        suffix = Path(safe_name).suffix.lower()
        if suffix not in _ALLOWED_SUFFIXES:
            raise ValueError("only .sh, .txt, or extensionless text files are allowed")

        try:
            text = content.decode("utf-8")
        except UnicodeDecodeError as exc:
            raise ValueError("only UTF-8 text files are allowed") from exc

        existing = {path.name for path in self._all_files()}
        if safe_name not in existing and len(existing) >= MAX_HOOK_FILES:
            raise ValueError("only the first 3 files are supported here")

        target = self.scripts_dir / safe_name
        target.write_text(text, encoding="utf-8")
        if os.name != "nt" and suffix == ".sh":
            current_mode = target.stat().st_mode
            target.chmod(current_mode | 0o700)
        return self._file_payload(target)

    def read_text_file(self, filename: str) -> dict[str, str]:
        path = self._resolve_file(filename)
        try:
            content = path.read_text(encoding="utf-8")
        except UnicodeDecodeError as exc:
            raise ValueError("this file cannot be viewed because it is not text") from exc
        return {"filename": path.name, "content": content}

    def delete_file(self, filename: str) -> None:
        path = self._resolve_file(filename)
        path.unlink(missing_ok=True)

    def run_command(self, command: str, *, phase: str, context: dict[str, Any]) -> None:
        normalized = str(command or "").strip()
        if not normalized:
            return

        shell_command = self._resolve_command(normalized)
        env = os.environ.copy()
        env.update(self._build_env(phase, context))

        try:
            result = subprocess.run(
                shell_command,
                shell=True,
                cwd=self.scripts_dir,
                env=env,
                capture_output=True,
                text=True,
                timeout=_HOOK_TIMEOUT_SECONDS,
            )
        except Exception as exc:
            self.logger.warning("hook_execution_failed phase=%s command=%s detail=%s", phase, normalized, exc)
            return

        if result.returncode != 0:
            detail = (result.stderr or result.stdout or "").strip()
            self.logger.warning(
                "hook_execution_nonzero phase=%s command=%s code=%s detail=%s",
                phase,
                normalized,
                result.returncode,
                detail,
            )

    def _all_files(self) -> list[Path]:
        self.scripts_dir.mkdir(parents=True, exist_ok=True)
        return sorted((path for path in self.scripts_dir.iterdir() if path.is_file()), key=lambda path: path.name.lower())

    def _resolve_command(self, command: str) -> str:
        stripped = command.strip()
        if any(char.isspace() for char in stripped):
            return stripped
        candidate = self.scripts_dir / stripped
        if candidate.exists() and candidate.is_file():
            prefix = ".\\" if os.name == "nt" else "./"
            return f"{prefix}{candidate.name}"
        return stripped

    def _build_env(self, phase: str, context: dict[str, Any]) -> dict[str, str]:
        env = {
            "RELAY_APP": "central",
            "RELAY_HOOK_PHASE": phase,
            "RELAY_HOOK_SCRIPTS_DIR": str(self.scripts_dir),
        }
        for key, value in context.items():
            env_key = f"RELAY_{key.upper()}"
            env[env_key] = "" if value is None else str(value)
        return env

    def _resolve_file(self, filename: str) -> Path:
        safe_name = self._sanitize_filename(filename)
        path = self.scripts_dir / safe_name
        if not path.exists() or not path.is_file():
            raise FileNotFoundError(safe_name)
        return path

    def _sanitize_filename(self, filename: str) -> str:
        safe_name = Path(str(filename or "").strip()).name
        if not safe_name or safe_name in {".", ".."}:
            raise ValueError("filename is required")
        return safe_name

    def _file_payload(self, path: Path) -> dict[str, Any]:
        try:
            viewable = True
            path.read_text(encoding="utf-8")
        except UnicodeDecodeError:
            viewable = False

        return {
            "name": path.name,
            "size_bytes": path.stat().st_size,
            "modified_at": datetime.fromtimestamp(path.stat().st_mtime, tz=timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
            "viewable": viewable,
        }

File: central/internal/services/hooks.go
package services

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
	"unicode/utf8"
)

const (
	MaxHookFiles       = 3
	hookTimeoutSeconds = 300
)

var allowedHookSuffixes = map[string]bool{".sh": true, ".txt": true}

type HookManager struct {
	ScriptsDir string
	logger     *slog.Logger
}

func NewHookManager(scriptsDir string, logger *slog.Logger) *HookManager {
	os.MkdirAll(scriptsDir, 0o755)
	return &HookManager{ScriptsDir: scriptsDir, logger: logger}
}

type HookFileInfo struct {
	Name       string `json:"name"`
	SizeBytes  int64  `json:"size_bytes"`
	ModifiedAt string `json:"modified_at"`
	Viewable   bool   `json:"viewable"`
}

func (h *HookManager) Snapshot(preCommand, postCommand string) map[string]any {
	return map[string]any{
		"pre_command":  preCommand,
		"post_command": postCommand,
		"script_dir":   h.ScriptsDir,
		"max_files":    MaxHookFiles,
		"files":        h.ListFiles(),
	}
}

func (h *HookManager) ListFiles() []HookFileInfo {
	os.MkdirAll(h.ScriptsDir, 0o755)
	entries, _ := os.ReadDir(h.ScriptsDir)
	var files []HookFileInfo
	for _, e := range entries {
		if e.IsDir() || len(files) >= MaxHookFiles {
			continue
		}
		info, err := e.Info()
		if err != nil {
			continue
		}
		files = append(files, hookFileInfo(filepath.Join(h.ScriptsDir, e.Name()), info))
	}
	return files
}

func (h *HookManager) SaveUploadedFile(filename string, content []byte) (HookFileInfo, error) {
	safeName := sanitizeFilename(filename)
	if safeName == "" {
		return HookFileInfo{}, fmt.Errorf("filename is required")
	}
	ext := strings.ToLower(filepath.Ext(safeName))
	if !allowedHookSuffixes[ext] {
		return HookFileInfo{}, fmt.Errorf("only .sh scripts or .txt helper files are allowed")
	}
	text := string(content)
	if !isUTF8(content) {
		return HookFileInfo{}, fmt.Errorf("only UTF-8 text files are allowed")
	}

	existing := h.existingNames()
	if _, known := existing[safeName]; !known && len(existing) >= MaxHookFiles {
		return HookFileInfo{}, fmt.Errorf("only the first 3 files are supported here")
	}

	if ext == ".sh" {
		text = strings.ReplaceAll(strings.ReplaceAll(text, "\r\n", "\n"), "\r", "\n")
	}

	target := filepath.Join(h.ScriptsDir, safeName)
	if err := os.WriteFile(target, []byte(text), 0o644); err != nil {
		return HookFileInfo{}, err
	}
	if ext == ".sh" {
		info, _ := os.Stat(target)
		if info != nil {
			os.Chmod(target, info.Mode()|0o700)
		}
	}
	info, err := os.Stat(target)
	if err != nil {
		return HookFileInfo{}, err
	}
	return hookFileInfo(target, info), nil
}

func (h *HookManager) ReadTextFile(filename string) (string, string, error) {
	safeName := sanitizeFilename(filename)
	path := filepath.Join(h.ScriptsDir, safeName)
	content, err := os.ReadFile(path)
	if os.IsNotExist(err) {
		return "", "", fmt.Errorf("%s: not found", safeName)
	}
	if err != nil {
		return "", "", err
	}
	if !isUTF8(content) {
		return "", "", fmt.Errorf("this file cannot be viewed because it is not text")
	}
	return safeName, string(content), nil
}

func (h *HookManager) DeleteFile(filename string) error {
	safeName := sanitizeFilename(filename)
	path := filepath.Join(h.ScriptsDir, safeName)
	err := os.Remove(path)
	if os.IsNotExist(err) {
		return fmt.Errorf("%s: not found", safeName)
	}
	return err
}

func (h *HookManager) RunCommand(command, phase string, hookCtx map[string]any) {
	normalized := strings.TrimSpace(command)
	if normalized == "" {
		return
	}

	shellCmd := h.resolveCommand(normalized)
	env := os.Environ()
	env = append3to1goEnv(env, "APP", "central")
	env = append3to1goEnv(env, "HOOK_PHASE", phase)
	env = append3to1goEnv(env, "HOOK_SCRIPTS_DIR", h.ScriptsDir)
	for k, v := range hookCtx {
		val := ""
		if v != nil {
			val = fmt.Sprintf("%v", v)
		}
		env = append3to1goEnv(env, strings.ToUpper(k), val)
	}

	ctx, cancel := context.WithTimeout(context.Background(), hookTimeoutSeconds*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "sh", "-c", shellCmd)
	cmd.Dir = h.ScriptsDir
	cmd.Env = env

	stdout, runErr := cmd.Output()
	var stderr []byte
	if exitErr, ok := runErr.(*exec.ExitError); ok {
		stderr = exitErr.Stderr
	}
	if ctx.Err() == context.DeadlineExceeded {
		h.logger.Warn("hook_execution_timeout", "phase", phase, "command", normalized)
		return
	}

	if len(stdout) > 0 {
		h.logger.Info("hook_execution_stdout", "phase", phase, "command", normalized, "output", strings.TrimSpace(string(stdout)))
	}
	if len(stderr) > 0 {
		h.logger.Warn("hook_execution_stderr", "phase", phase, "command", normalized, "output", strings.TrimSpace(string(stderr)))
	}
	if runErr != nil {
		h.logger.Warn("hook_execution_nonzero", "phase", phase, "command", normalized, "error", runErr)
	}
}

func append3to1goEnv(env []string, name, value string) []string {
	return append(env, "THREETOONEGO_"+name+"="+value)
}

func (h *HookManager) resolveCommand(command string) string {
	if strings.ContainsAny(command, " \t") {
		return command
	}
	candidate := filepath.Join(h.ScriptsDir, command)
	if info, err := os.Stat(candidate); err == nil && !info.IsDir() {
		return "./" + command
	}
	return command
}

func (h *HookManager) existingNames() map[string]struct{} {
	entries, _ := os.ReadDir(h.ScriptsDir)
	m := make(map[string]struct{})
	for _, e := range entries {
		if !e.IsDir() {
			m[e.Name()] = struct{}{}
		}
	}
	return m
}

func hookFileInfo(path string, info os.FileInfo) HookFileInfo {
	content, err := os.ReadFile(path)
	viewable := err == nil && isUTF8(content)
	return HookFileInfo{
		Name:       info.Name(),
		SizeBytes:  info.Size(),
		ModifiedAt: info.ModTime().UTC().Format("2006-01-02T15:04:05Z"),
		Viewable:   viewable,
	}
}

func sanitizeFilename(filename string) string {
	base := filepath.Base(strings.TrimSpace(filename))
	if base == "." || base == ".." {
		return ""
	}
	return base
}

func isUTF8(data []byte) bool {
	return utf8.Valid(data)
}
